File: leetcode/go/4 Median of Two Sorted Arrays.go
func findKth(nums1 []int, nums2 []int, i int, j int, k int) float64 {
    if(i >= len(nums1)){
        return float64(nums2[j + k - 1])
    }
    if(j >= len(nums2)){
        return float64(nums1[i + k - 1])
    }
    
    if(k == 1){
        if(nums1[i] < nums2[j]){
            return float64(nums1[i])
        }else{
            return float64(nums2[j])
        }
    }
    
    midVal1 := 9223372036854775807
    midVal2 := 9223372036854775807
    
    if(i + k / 2 - 1 < len(nums1)){
        midVal1 = nums1[i + k / 2 - 1]
    }
    
    if(j + k / 2 - 1 < len(nums2)){
        midVal2 = nums2[j + k / 2 - 1]
    }
    
    if(midVal1 < midVal2){
        return findKth(nums1, nums2, i + k / 2, j, k - k / 2)
    }else{
        return findKth(nums1, nums2, i, j + k / 2, k - k / 2)
    }
}

func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
    l1 := len(nums1)
    l2 := len(nums2)
    
    left := (l1 + l2 + 1) / 2
    right := (l1 + l2 + 2) / 2
    
    return (findKth(nums1, nums2, 0, 0, left) + findKth(nums1, nums2, 0, 0, right)) / 2
}

File: leetcode/ts/4 Median of Two Sorted Arrays.ts
function findKth(nums1 : number[], i : number, nums2 : number[], j : number, target : number) : number{
    if(i >= nums1.length){
        return nums2[j + target - 1]
    }
    if(j >= nums2.length){
        return nums1[i + target - 1]
    }
    if(target == 1){
        return Math.min(nums1[i], nums2[j])
    }
    let midVal1 = (i + Math.floor(target / 2) - 1 < nums1.length) ? nums1[i + Math.floor(target / 2) - 1] : Number.MAX_VALUE
    let midVal2 = (j + Math.floor(target / 2) - 1 < nums2.length) ? nums2[j + Math.floor(target / 2) - 1] : Number.MAX_VALUE
    
    if(midVal1 < midVal2){
        return findKth(nums1, i + Math.floor(target / 2), nums2, j, target - Math.floor(target / 2))
    }else{
        return findKth(nums1, i, nums2, j + Math.floor(target / 2), target - Math.floor(target / 2))
    }
}

function findMedianSortedArrays(nums1: number[], nums2: number[]): number {
    const len1 = nums1.length
    const len2 = nums2.length
    let k1 = Math.floor((len1 + len2 + 1) / 2)
    let k2 = Math.floor((len1 + len2 + 2) / 2)
    
    return (findKth(nums1, 0, nums2, 0, k1) + findKth(nums1, 0, nums2, 0, k2)) / 2
};